In [1]:
import pandas as pd

In [2]:
import regex as re

In [4]:
titanic_db = pd.read_csv('../Data/train.csv')


In [5]:
titanic_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [6]:
titanic_db.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [7]:
change_db=titanic_db.copy(deep=True)


In [8]:
change_db.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [9]:
change_db.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [10]:
pattern = (
    r'^(?P<LastName>[^,]+),\s*'
    r'(?P<Title>\w+)\.?\s*'
    r'(?P<HusbandName>[^(]+)?'
    r'(?:\((?P<ActualName>[^)]+)\))?'
)

extract= change_db['Name'].str.extract(pattern)
change_db2=pd.concat([change_db,extract],axis=1)
change_db2.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked   LastName Title  \
0      0         A/5 21171   7.2500   NaN        S     Braund    Mr   
1      0          PC 17599  71.2833   C85        C    Cumings   Mrs   
2      0  STON/O2. 3101282   7.9250   NaN        S  Heikkinen  Miss   
3      0            113803  53.1000  C123        S   Futrelle   Mrs   
4      0            373450   8.0500   NaN        S      Allen    Mr   

      HusbandName              ActualName  
0     Owen Harris                     NaN  
1   John Bradley   Florence Briggs Thayer  
2           Laina                     NaN  
3  Jacques Heath            Lily May Peel  
4   William Henry                     NaN

In [11]:
change_db2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
 12  LastName     891 non-null    object 
 13  Title        891 non-null    object 
 14  HusbandName  875 non-null    object 
 15  ActualName   143 non-null    object 
dtypes: float64(2), int64(5), object(9)
memory usage: 111.5+ KB


In [12]:
change_db2.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
LastName         0
Title            0
HusbandName     16
ActualName     748
dtype: int64

In [12]:
change_db2.loc[change_db2['HusbandName'].isnull()]

PassengerId  Survived  Pclass  \
15            16         1       2   
66            67         1       2   
166          167         1       1   
190          191         1       2   
259          260         1       2   
272          273         1       2   
327          328         1       2   
362          363         0       3   
367          368         1       3   
483          484         1       3   
516          517         1       2   
556          557         1       1   
564          565         0       3   
772          773         0       2   
823          824         1       3   
865          866         1       2   

                                                  Name     Sex   Age  SibSp  \
15                    Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
66                        Nye, Mrs. (Elizabeth Ramell)  female  29.0      0   
166             Chibnall, Mrs. (Edith Martha Bowerman)  female   NaN      0   
190                                Pinsky, Mrs. (Rosa)  female  32.0      0   
259                        Parrish, Mrs. (Lutie Davis)  female  50.0      0   
272          Mellinger, Mrs. (Elizabeth Anne Maidment)  female  41.0      0   
327                            Ball, Mrs. (Ada E Hall)  female  36.0      0   
362                    Barbara, Mrs. (Catherine David)  female  45.0      0   
367                     Moussa, Mrs. (Mantoura Boulos)  female   NaN      0   
483                             Turkula, Mrs. (Hedwig)  female  63.0      0   
516                       Lemore, Mrs. (Amelia Milley)  female  34.0      0   
556  Duff Gordon, Lady. (Lucille Christiana Sutherl...  female  48.0      1   
564                     Meanwell, Miss. (Marion Ogden)  female   NaN      0   
772                                  Mack, Mrs. (Mary)  female  57.0      0   
823                                 Moor, Mrs. (Beila)  female  27.0      0   
865                           Bystrom, Mrs. (Karolina)  female  42.0      0   

     Parch             Ticket     Fare Cabin Embarked     LastName Title  \
15       0             248706  16.0000   NaN        S      Hewlett   Mrs   
66       0         C.A. 29395  10.5000   F33        S          Nye   Mrs   
166      1             113505  55.0000   E33        S     Chibnall   Mrs   
190      0             234604  13.0000   NaN        S       Pinsky   Mrs   
259      1             230433  26.0000   NaN        S      Parrish   Mrs   
272      1             250644  19.5000   NaN        S    Mellinger   Mrs   
327      0              28551  13.0000     D        S         Ball   Mrs   
362      1               2691  14.4542   NaN        C      Barbara   Mrs   
367      0               2626   7.2292   NaN        C       Moussa   Mrs   
483      0               4134   9.5875   NaN        S      Turkula   Mrs   
516      0         C.A. 34260  10.5000   F33        S       Lemore   Mrs   
556      0              11755  39.6000   A16        C  Duff Gordon  Lady   
564      0  SOTON/O.Q. 392087   8.0500   NaN        S     Meanwell  Miss   
772      0        S.O./P.P. 3  10.5000   E77        S         Mack   Mrs   
823      1             392096  12.4750  E121        S         Moor   Mrs   
865      0             236852  13.0000   NaN        S      Bystrom   Mrs   

    HusbandName                     ActualName  
15          NaN                Mary D Kingcome  
66          NaN               Elizabeth Ramell  
166         NaN          Edith Martha Bowerman  
190         NaN                           Rosa  
259         NaN                    Lutie Davis  
272         NaN        Elizabeth Anne Maidment  
327         NaN                     Ada E Hall  
362         NaN                Catherine David  
367         NaN                Mantoura Boulos  
483         NaN                         Hedwig  
516         NaN                  Amelia Milley  
556         NaN  Lucille Christiana Sutherland  
564         NaN                   Marion Ogden  
772         NaN                    

In [13]:
change_db2['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [14]:
change_db2['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [15]:
change_db2.loc[change_db2['Survived']==1, 'Sex'].value_counts()

Sex
female    233
male      109
Name: count, dtype: int64

In [16]:
change_db2.loc[((change_db2['Age'].isnull())),'Sex'].value_counts()

Sex
male      124
female     53
Name: count, dtype: int64

In [14]:
change_db2.loc[((change_db2['Age'].isnull())),'Pclass'].value_counts()

Pclass
3    136
1     30
2     11
Name: count, dtype: int64

In [15]:
change_db2.loc[((change_db2['Age'].isnull())),'Title'].value_counts()

Title
Mr        119
Miss       36
Mrs        17
Master      4
Dr          1
Name: count, dtype: int64

In [16]:
change_db2.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Dr'))]

PassengerId  Survived  Pclass                       Name   Sex  Age  \
766          767         0       1  Brewe, Dr. Arthur Jackson  male  NaN   

     SibSp  Parch  Ticket  Fare Cabin Embarked LastName Title     HusbandName  \
766      0      0  112379  39.6   NaN        C    Brewe    Dr  Arthur Jackson   

    ActualName  
766        NaN

In [17]:
change_db2.loc[((change_db2['Title']=='Dr'))]

PassengerId  Survived  Pclass                           Name     Sex  \
245          246         0       1    Minahan, Dr. William Edward    male   
317          318         0       2           Moraweck, Dr. Ernest    male   
398          399         0       2               Pain, Dr. Alfred    male   
632          633         1       1      Stahelin-Maeglin, Dr. Max    male   
660          661         1       1  Frauenthal, Dr. Henry William    male   
766          767         0       1      Brewe, Dr. Arthur Jackson    male   
796          797         1       1    Leader, Dr. Alice (Farnham)  female   

      Age  SibSp  Parch    Ticket      Fare Cabin Embarked          LastName  \
245  44.0      2      0     19928   90.0000   C78        Q           Minahan   
317  54.0      0      0     29011   14.0000   NaN        S          Moraweck   
398  23.0      0      0    244278   10.5000   NaN        S              Pain   
632  32.0      0      0     13214   30.5000   B50        C  Stahelin-Maeglin   
660  50.0      2      0  PC 17611  133.6500   NaN        S        Frauenthal   
766   NaN      0      0    112379   39.6000   NaN        C             Brewe   
796  49.0      0      0     17465   25.9292   D17        S            Leader   

    Title     HusbandName ActualName  
245    Dr  William Edward        NaN  
317    Dr          Ernest        NaN  
398    Dr          Alfred        NaN  
632    Dr             Max        NaN  
660    Dr   Henry William        NaN  
766    Dr  Arthur Jackson        NaN  
796    Dr          Alice     Farnham

There are 7 `Dr.` passengers, where **1** is *female* and 6 are *male* Dr. passengers. 


In [22]:
change_db2.loc[((change_db2['Sex']=='male')&(change_db2['Title']=='Dr'))].describe()

PassengerId  Survived    Pclass        Age     SibSp  Parch        Fare
count     6.000000  6.000000  6.000000   5.000000  6.000000    6.0    6.000000
mean    504.000000  0.333333  1.333333  40.600000  0.666667    0.0   53.041667
std     211.018483  0.516398  0.516398  12.876335  1.032796    0.0   48.757506
min     246.000000  0.000000  1.000000  23.000000  0.000000    0.0   10.500000
25%     338.250000  0.000000  1.000000  32.000000  0.000000    0.0   18.125000
50%     516.000000  0.000000  1.000000  44.000000  0.000000    0.0   35.050000
75%     654.000000  0.750000  1.750000  50.000000  1.500000    0.0   77.400000
max     767.000000  1.000000  2.000000  54.000000  2.000000    0.0  133.650000

In [19]:
change_db_cleaned=change_db2.copy(deep=True)

For final cleaning created a table `change_db_cleaned` by creating a copy of `change_db2`

In [20]:
change_db_cleaned.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Dr')),'Age']=41

In [21]:
change_db_cleaned.loc[((change_db2['Sex']=='male')&(change_db2['Title']=='Dr'))].describe()

PassengerId  Survived    Pclass        Age     SibSp  Parch        Fare
count     6.000000  6.000000  6.000000   6.000000  6.000000    6.0    6.000000
mean    504.000000  0.333333  1.333333  40.666667  0.666667    0.0   53.041667
std     211.018483  0.516398  0.516398  11.518102  1.032796    0.0   48.757506
min     246.000000  0.000000  1.000000  23.000000  0.000000    0.0   10.500000
25%     338.250000  0.000000  1.000000  34.250000  0.000000    0.0   18.125000
50%     516.000000  0.000000  1.000000  42.500000  0.000000    0.0   35.050000
75%     654.000000  0.750000  1.750000  48.500000  1.500000    0.0   77.400000
max     767.000000  1.000000  2.000000  54.000000  2.000000    0.0  133.650000

`Dr.` title (1 null): Impute with **40.6** rounding it to **41**— mean age of male Dr. passengers.

In [23]:
change_db2.loc[((change_db2['Age'].isnull())&(change_db2['Title']=='Master'))]

PassengerId  Survived  Pclass  \
65            66         1       3   
159          160         0       3   
176          177         0       3   
709          710         1       3   

                                                  Name   Sex  Age  SibSp  \
65                            Moubarek, Master. Gerios  male  NaN      1   
159                         Sage, Master. Thomas Henry  male  NaN      8   
176                      Lefebre, Master. Henry Forbes  male  NaN      3   
709  Moubarek, Master. Halim Gonios ("William George")  male  NaN      1   

     Parch    Ticket     Fare Cabin Embarked  LastName   Title    HusbandName  \
65       1      2661  15.2458   NaN        C  Moubarek  Master         Gerios   
159      2  CA. 2343  69.5500   NaN        S      Sage  Master   Thomas Henry   
176      1      4133  25.4667   NaN        S   Lefebre  Master   Henry Forbes   
709      1      2661  15.2458   NaN        C  Moubarek  Master  Halim Gonios    

           ActualName  
65                NaN  
159               NaN  
176               NaN  
709  "William George"

`Master` title of 3rd class has all 4 null values of age

In [28]:
change_db2.loc[((change_db2['Title']=='Master')&(change_db2['Pclass']==3))].describe()

PassengerId   Survived  Pclass        Age      SibSp      Parch  \
count    28.000000  28.000000    28.0  24.000000  28.000000  28.000000   
mean    388.285714   0.392857     3.0   5.350833   2.928571   1.357143   
std     314.309927   0.497347     0.0   3.593608   1.942289   0.558721   
min       8.000000   0.000000     3.0   0.420000   0.000000   0.000000   
25%     151.500000   0.000000     3.0   2.000000   1.000000   1.000000   
50%     270.500000   0.000000     3.0   4.000000   3.500000   1.000000   
75%     761.000000   1.000000     3.0   9.000000   4.000000   2.000000   
max     870.000000   1.000000     3.0  12.000000   8.000000   2.000000   

            Fare  
count  28.000000  
mean   28.177232  
std    13.898033  
min     8.516700  
25%    15.900000  
50%    28.512500  
75%    33.462500  
max    69.550000

In [31]:
change_db_cleaned.loc[(change_db2['Title']=='Master')&(change_db2['Age'].isnull()),'Age']=5

In [32]:
change_db_cleaned.loc[((change_db2['Title']=='Master')&(change_db2['Pclass']==3))].describe()

PassengerId   Survived  Pclass        Age      SibSp      Parch  \
count    28.000000  28.000000    28.0  28.000000  28.000000  28.000000   
mean    388.285714   0.392857     3.0   5.300714   2.928571   1.357143   
std     314.309927   0.497347     0.0   3.319105   1.942289   0.558721   
min       8.000000   0.000000     3.0   0.420000   0.000000   0.000000   
25%     151.500000   0.000000     3.0   2.750000   1.000000   1.000000   
50%     270.500000   0.000000     3.0   5.000000   3.500000   1.000000   
75%     761.000000   1.000000     3.0   8.250000   4.000000   2.000000   
max     870.000000   1.000000     3.0  12.000000   8.000000   2.000000   

            Fare  
count  28.000000  
mean   28.177232  
std    13.898033  
min     8.516700  
25%    15.900000  
50%    28.512500  
75%    33.462500  
max    69.550000

`Master` in 3rd class (4 nulls): Impute with **5.3** round off to **5** — mean age of Master passengers in 3rd class.


## Final Age Imputation Decisions

- `Dr.` title (1 null): Impute with **40.6** — mean age of male Dr. passengers.
- `Master` in 3rd class (4 nulls): Impute with **5.3** — mean age of Master passengers in 3rd class.
